### PR TITLE
compas_viewer scene object

### PR DESCRIPTION
**Pull request by ZacZhangzhuo**
see https://github.com/compas-dev/compas_viewer/pull/74

## Major Changes
* **Added `RobotModelObject` and its example in the documentation**. the `RobotModelObject` class is the main focus of this PR.

## Test
Example file can be found at : https://compas.dev/compas_viewer/latest/examples/object/robot.html

### CODE REVIEW

**Review by tomvanmele — APPROVED** [review]

**Comment by jf---** [comment]
@gonzalocasas or @tomvanmele anything keeping you back from merging? cool contribution @ZacZhangzhuo

File: CHANGELOG.md [hunk 1]
```diff
@@ -8,7 +8,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ## Unreleased
 
 ### Added
-
+* Added `compas_viewer`'s scene object.
```
Comment (gonzalocasas) [inline CHANGELOG.md:11]:
```suggestion
* Added support for `compas_viewer`

```

File: src/compas_robots/viewer/scene/robotmodelobject.py [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+        return self._show_visual
+
+    @show_visual.setter
+    def show_visual(self, value: bool):
+        if value == self._show_visual:
+            return
+        self._show_visual = value
+        for i, visual_object in enumerate(self.visual_objects):
+            if value:
+                self.scene.tree.add_object(visual_object, self.visual_objects[i - 1] if i > 0 else self)
```
Comment (gonzalocasas) [inline src/compas_robots/viewer/scene/robotmodelobject.py:72]: I think the conditional inside the method call is hard to read, please assign the value of that to a variable and then just pass it to `add_object`

File: src/compas_robots/viewer/scene/robotmodelobject.py [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+        return self._show_collision
+
+    @show_collision.setter
+    def show_collision(self, value: bool):
+        if value == self._show_collision:
+            return
+        self._show_collision = value
+        for i, collision_object in enumerate(self.collision_objects):
+            if value:
+                self.scene.tree.add_object(collision_object, self.visual_objects[i - 1] if i > 0 else self)
```
Comment (gonzalocasas) [inline src/compas_robots/viewer/scene/robotmodelobject.py:88]: same

**Comment by gonzalocasas** [comment]
> @gonzalocasas or @tomvanmele anything keeping you back from merging? cool contribution @ZacZhangzhuo

thanks for the poke!

**Review by gonzalocasas — APPROVED** [review]
LGTM! I added some minor feedback but overall it's looking great! I re-triggered the build: after it goes green and you address the comments below, we can merge!

File: src/compas_robots/viewer/scene/__init__.py [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+from compas.scene import register
+
+from compas_robots import RobotModel
+
+from .robotmodelobject import RobotModelObject
+
+
+@plugin(category="factories", requires=["compas_viewer"])
+def register_scene_objects():
+    register(RobotModel, RobotModelObject, context="Viewer")
```
Comment (gonzalocasas) [inline src/compas_robots/viewer/scene/__init__.py:11]: Correct me if I'm wrong, but I think the build fails because it's requiring `compas_viewer` to be installed, which shouldn't be the case. If you move the import of `from .robotmodelobject import RobotModelObject` inside the `register_scene_objects` function, it will not try to import if the `requires` is not fulfilled.

**Comment by ZacZhangzhuo** [comment]
> @gonzalocasas or @tomvanmele anything keeping you back from merging? cool contribution @ZacZhangzhuo

thanks! :-)

File: src/compas_robots/viewer/scene/robotmodelobject.py [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+                self.scene.instance_colors[collision_object.instance_color.rgb255] = collision_object
+            else:
+                self.scene.remove(collision_object)
+
+    def init(self):
+        """Initialize the robot object with creating the visual and collision objects."""
+        for i, visual_object in enumerate(self.visual_objects):
+            visual_object.init()
+            if self.show_visual:
+                self.scene.add(visual_object, self.visual_objects[i - 1] if i > 0 else self)
```
Comment (gonzalocasas) [inline src/compas_robots/viewer/scene/robotmodelobject.py:104]: Could you please add the same simplification of the condition here?

File: src/compas_robots/viewer/scene/robotmodelobject.py [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+                self.scene.instance_colors[visual_object.instance_color.rgb255] = visual_object
+
+        for i, collision_object in enumerate(self.collision_objects):
+            collision_object.init()
+            if self.show_collision:
+                self.scene.add(collision_object, self.visual_objects[i - 1] if i > 0 else self)
```
Comment (gonzalocasas) [inline src/compas_robots/viewer/scene/robotmodelobject.py:110]: Could you please add the same simplification of the condition also here?

**Comment by ZacZhangzhuo** [comment]
the build still failed even with:
```python

@plugin(category="factories", requires=["compas_viewer"])
def register_scene_objects():
    from .robotmodelobject import RobotModelObject

    register(RobotModel, RobotModelObject, context="Viewer")

```
@gonzalocasas

**Comment by gonzalocasas** [comment]
Sorry, I just realized that something is off. And it's that we don't need to have the viewer plugin inside `compas_robots`, that's exactly why these things are plugins, so that this module can live inside the repo of `compas_viewer` and be auto-discovered

**Comment by ZacZhangzhuo** [comment]
> Sorry, I just realized that something is off. And it's that we don't need to have the viewer plugin inside `compas_robots`, that's exactly why these things are plugins, so that this module can live inside the repo of `compas_viewer` and be auto-discovered

didn't really get it. How should we modify the code then? i created this __init__.py based on `src\compas_robots\rhino\scene\__init__.py`, as a template.

**Comment by gonzalocasas** [comment]
> > Sorry, I just realized that something is off. And it's that we don't need to have the viewer plugin inside `compas_robots`, that's exactly why these things are plugins, so that this module can live inside the repo of `compas_viewer` and be auto-discovered
> 
> didn't really get it. How should we modify the code then? i created this **init**.py based on `src\compas_robots\rhino\scene\__init__.py`, as a template.

The code is all correct, but I think it just needs to be moved to the `compas_viewer` project instead of here.

**Comment by tomvanmele** [comment]
@gonzalocasas is it not exactly the other way around?! since it is a plugin system, not everything needs to be implemented in the viewer (since the viewer can't know about all the packages that want to plug into it), but rather in the packages themselves? Rhino scene objects are also implemented per package and not all in `compas_rhino`, for the same reason...

**Comment by gonzalocasas** [comment]
Alright, this one was really confusing to reason about. I thought about it a bunch, then discussed it also with @yck011522.

So, in terms of dependencies, this is correct. The plugin part takes care of decoupling the dependency from `compas_robots` to `compas_viewer`, and at the same time, if the interface on the viewer side is correct, viewer should not be dependent on `compas_robots` either. So that's all good. 

In this diagram (directed dashed line represents a "A depends on B" relationship), it's highlighted how the sub-packages of this repository that deal with each CAD/viewer have dependencies on those things (Blender, Viewer and Rhino respectively), and also have a dependency on the `RobotModel` class of `compas_robots`:
image

My confusion came from the fact of what does a developer of different parts of the system need to be aware of when developing:
 - As a downside of merging this here, a developer of `compas_viewer` now needs to be aware that there are features that depend on the viewer that are spread across multiple repositories,
 - But on the positive side, a developer of `compas_robots` will immediately be aware that changes to the `RobotModel` class need to be propagated to all the implementations of scene objects, and it will all be in this repository.

We don't have this problem so clear in the case of Rhino/Blender, because those code bases are external to us, and none of us develops Rhino or Blender themselves, but viewer is a bit of a different case.

And here there's a bit of an assumption, but I expect the API of `compas_robots` and its `RobotModel` class to change more often than the base API that `compas_viewer` requires for it to be able to support a new object type, so, in that case, it makes sense that this stays in this repository.

Bottom line: we keep it here. The only thing remaining is to fix the build.

**Review by gonzalocasas — CHANGES_REQUESTED** [review]
Fix the build and we're good to go

**Comment by gonzalocasas** [comment]
This one has been replaced by https://github.com/compas-dev/compas_robots/pull/14
